mlflow/protos/databricks.proto

```diff
@@ -217,18 +330,82 @@ enum ErrorCode {
+  // The caller does not have permission to execute the specified operation.
+  // PERMISSION_DENIED must not be used for rejections caused by exhausting some resource,
+  // use RESOURCE_EXHAUSTED instead for those errors.
+  // PERMISSION_DENIED must not be used if the caller can not be identified,
+  // use CUSTOMER_UNAUTHORIZED instead for those errors.
+  // This error code does not imply the request is valid or the requested entity exists or
+  // satisfies other pre-conditions.
+  //
+  // Maps to:
+  // - google.rpc.Code: PERMISSION_DENIED = 7;
+  // - HTTP code: 403 Forbidden
   PERMISSION_DENIED = 1004;
 
-  // If a given user/entity is trying to use a feature which has been disabled
+  // If a given user/entity is trying to use a feature which has been disabled.
+  //
+  // Maps to:
+  // - google.rpc.Code: NOT_FOUND = 5;
+  // - HTTP code: 404 Not Found
   FEATURE_DISABLED = 1005;
 
-  // If customer-provided credentials are not authorized to perform an operation
+  // The request does not have valid authentication (AuthN) credentials for the operation.
+  //
+  // For authentication (AuthN) errors prefer using UNAUTHENTICATED, unless you need to keep
+  // consistent behavior with legacy code.
+  // For authorization (AuthZ) errors use PERMISSION_DENIED.
+  //
+  // Important: name is confusing, this error code is for authentication (AuthN) errors, not
+  // authorization (AuthZ) errors. It maps to 401 Unauthorized and suffers from the same confusing
+  // naming. See https://datatracker.ietf.org/doc/html/rfc7235#section-3.1 - "[...] status code
+  // indicates that the request has not been applied because it lacks valid authentication
+  // credentials for the target resource. [...] If the request included authentication credentials,
+  // then the 401 response indicates that authorization has been refused for those credentials."
+  //
+  // Also, see https://stackoverflow.com/a/6937030/16352922, it covers it pretty well.
+  //
+  // Maps to:
+  // - google.rpc.Code: UNAUTHENTICATED = 16;
+  // - HTTP code: 401 Unauthorized
   CUSTOMER_UNAUTHORIZED = 1006;
 
-  // If the API request is rejected due to throttling
+  // If the API request is rejected due to throttling.
+  // Prefer a more generic RESOURCE_EXHAUSTED for the new use cases.
+  //
+  // Maps to:
+  // - google.rpc.Code: RESOURCE_EXHAUSTED = 8;
+  // - HTTP code: 429 Too Many Requests
   REQUEST_LIMIT_EXCEEDED = 1007;
 
+  // Indicates API request was rejected due a conflict with an existing resource.
+  RESOURCE_CONFLICT = 1008;
+
+  // Indicates that the HTTP response cannot be correctly deserialized.
+  // This currently is only used in DUST test clients, and not by any real service code.
+  UNPARSEABLE_HTTP_ERROR = 1009;
+
+  // The operation is not implemented or is not supported/enabled in this service.
+  //
+  // Maps to:
+  // - google.rpc.Code: UNIMPLEMENTED = 12;
+  // - HTTP code: 501 Not Implemented
```

Comment:
Does MlflowException map this to 501? Can we do that if not?

Reply:
Updated. We also need to map the new errors, right?

Reply:
Added `DATA_LOSS`.